src/Database/LSMTree/Internal/Run/Index/Compact/Construction.hs

```diff
@@ -261,20 +263,25 @@ unsafeEnd mci@MCompactIndex{..} = do
 
     -- Only slice out a chunk if there are entries in the chunk
     mchunk <- if ix == 0 then pure Nothing else do
-      cPrimary        <- VU.unsafeFreeze . VUM.slice 0 ix =<< readSTRef mciPrimary
-      cClashes        <- VU.unsafeFreeze . VUM.slice 0 ix =<< readSTRef mciClashes
-      cLargerThanPage <- VU.unsafeFreeze . VUM.slice 0 ix =<< readSTRef mciLargerThanPage
-      pure $ Just Chunk{..}
+      Just . Chunk <$> (VU.unsafeFreeze . VUM.slice 0 ix =<< readSTRef mciPrimary)
 
     -- We are not guaranteed to have seen all possible range-finder bit
     -- combinations, so we have to fill in the remainder of the rangerfinder
     -- vector.
     fillRangeFinderToEnd mci
     fcRangeFinder <- VU.unsafeFreeze mciRangeFinder
     let fcRangeFinderPrecision = mciRangeFinderPrecision
+    fcClashes <- fmap (VU.concat . reverse) $
+      traverse VU.unsafeFreeze . sliceCurrent ix =<< readSTRef mciClashes
     fcTieBreaker <- readSTRef mciTieBreaker
+    fcLargerThanPage <- fmap (VU.concat . reverse) $
+      traverse VU.unsafeFreeze . sliceCurrent ix =<< readSTRef mciLargerThanPage
 
     pure (mchunk, FinalChunk {..})
+  where
+    sliceCurrent ix (c NE.:| cs)
+      | ix == 0 = cs
+      | otherwise = VUM.slice 0 ix c : cs
```

Comment:
Could you give some insight into what this function does in a comment?

Reply:
done